aws/lambda/log-classifier/src/engine.rs

```diff
@@ -44,3 +45,381 @@ pub fn evaluate_ruleset(ruleset: &RuleSet, log: &Log) -> Option<Match> {
         .flat_map(|rule| evaluate_rule(rule, log))
         .max_by(|a, b| a.rule.priority.cmp(&b.rule.priority))
 }
+
+/// Evaluates a ruleset against all lines of a log, returning matching line numbers.
+///
+/// # Arguments
+///
+/// * `ruleset` - A reference to the RuleSet to evaluate.
+/// * `log` - A reference to the Log to evaluate against.
+///
+/// # Returns
+///
+/// A vector of line numbers that match any rule in the ruleset.
+/// Note: Each rule in ruleset can only match 1 line which is the last line that matches the rule.
+pub fn evaluate_ruleset_all_lines(ruleset: &RuleSet, log: &Log) -> Vec<usize> {
+    let matches: Vec<Match> = ruleset
+        .rules
+        .par_iter()
+        .flat_map(|rule| evaluate_rule(rule, log))
+        .collect();
+    // combine captures into a single vector
+    let mut lines = Vec::new();
+    for m in matches {
+        lines.push(m.line_number);
+    }
+    lines
+}
+
+/// Merges line numbers into chunks based on specified size constraints.
+///
+/// # Arguments
+///
+/// * `log` - A reference to the Log structure containing the lines.
+/// * `line_numbers` - A vector of line numbers to be chunked.
+/// * `min_context_padding` - The minimum size of each chunk.
+/// * `max_chunk_size` - The maximum size of each chunk.
```

Comment:
Sorry for the late review! I'm curious what is the use for `max_chunk_size` here as it makes the logic of this function more complicated.  What would be the drawback of just using `min_context_padding` around the line number (accepting the possibility of overlapping chunks)?

Reply:
>  I'm curious what is the use for max_chunk_size here as it makes the logic of this function more complicated.

There is a max context length on prompts to llms. The max_chunk_size is to enforce that we don't overflow that context length.


>  What would be the drawback of just using min_context_padding around the line number (accepting the possibility of overlapping chunks)? I guess more requests to Bedrock?

Yes that's right, but bedrock calls are cheap so it doesn't really matter.